### PR TITLE
[Feature] Hide category

### PR DESCRIPTION
**Pull request by Miodec**
- Added the ability to hide category groups.
 - Added a menu to the budget header which for now just has one option: toggle visibility of hidden categories
image
image
image

### CODE REVIEW

**Comment by MatissJanis** [comment]
:wave: Very cool!

Personally I've been using a "hidden categories" group where I placed my unused categories. But that's more of a workaround than a proper solution.. your solution is definitely better.

Before I do a proper code review - two bugs/issues spotted:

1. if you hide all the categories within a group - the group stays visible, but it has no categories; would it make sense to hive the group too then? I think that makes sense
2. the "toggle hidden categories" option is not persisted - refreshing the page hides all the categories; can we persist it to user prefs?

**Comment by Miodec** [comment]
> 👋 Very cool!
> 
> Personally I've been using a "hidden categories" group where I placed my unused categories. But that's more of a workaround than a proper solution.. your solution is definitely better.
> 
> Before I do a proper code review - two bugs/issues spotted:
> 
> 1. if you hide all the categories within a group - the group stays visible, but it has no categories; would it make sense to hive the group too then? I think that makes sense
> 2. the "toggle hidden categories" option is not persisted - refreshing the page hides all the categories; can we persist it to user prefs?

1. Oh yeah i forgot to point that out - i had a thought that the user might want to still see the group even if every category was hidden in case they wanted to add another category to that group. And also it could be a bit confusing if the category group dissapears, and the user might end up creating a copy? Just something to consider. Let me know, I can make it so the empty groups will not be shown.
2. Sure thing, could you point me to a component that already uses user prefs? Still learning the structure.

**Comment by j-f1** [comment]
> Sure thing, could you point me to a component that already uses user prefs? Still learning the structure.

The sidebar could be a good example of this since it preserves the pinned/floating state.

**Comment by shall0pass** [comment]
I'm loving this.  A couple of follow up items when this gets merged.
1. The set budget options to "Last months budget" and "3 month average" should probably only apply to visible categories.
2. Goal templates should probably only apply to visible categories.

It would be easy to lose money into a hidden category and not realize it if they're auto filled with those types of shortcuts.

**Comment by youngcw** [comment]
I like this.  Would it be doable to modify the nynab importer to automatically set the ynab hidden categories as hidden for actual?  I checked a ynab export and the hidden categories are labeled as such.

**Comment by Miodec** [comment]
Storing state in user prefs.
Added group hiding which is separate from category hiding.

Should be fully ready for review.

File: packages/desktop-client/src/components/budget/misc.js [hunk 1]
```diff
@@ -352,12 +383,15 @@ export function SidebarCategory({
...
-                } else {
+                } else if (type === 'delete') {
                   onDelete(category.id);
+                } else if (type === 'toggleHide') {
+                  onSave({ ...category, hidden: !category.hidden });
                 }
                 setMenuOpen(false);
               }}
               items={[
+                { name: 'toggleHide', text: category.hidden ? 'Show' : 'Hide' },
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/budget/misc.js:394]: **💬 suggestion:** ‏
```suggestion
                { name: 'toggleVisibility', text: category.hidden ? 'Show' : 'Hide' },
```

File: packages/desktop-client/src/components/budget/misc.js [hunk 2]
```diff
@@ -513,13 +547,19 @@ export function SidebarGroup({
...
                       onDelete(group.id);
+                    } else if (type === 'toggleHidden') {
+                      onSave({ ...group, hidden: !group.hidden });
                     }
                     setMenuOpen(false);
                   }}
                   items={[
                     { name: 'add-category', text: 'Add category' },
+                    {
+                      name: 'toggleHidden',
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/budget/misc.js:560]: **💬 suggestion:** ‏
```suggestion
                      name: 'toggleVisibility',
```

File: packages/desktop-client/src/components/budget/misc.js [hunk 3]
```diff
@@ -943,15 +1039,28 @@ const BudgetCategories = memo(
...
+          }
+
+          const groupCategories = group.categories.filter(
+            cat => showHiddenCategories || !cat.hidden,
+          );
+
+          // //checking group.categories.length to make sure groups with no categories still show
+          // if (group.categories.length > 0 && groupCategories.length === 0) {
+          //   return [];
+          // }
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/budget/misc.js:1053]: **💬 suggestion:** ‏can be removed

**Review by MatissJanis — CHANGES_REQUESTED** [review]
Sweet!

A few small comments, but apart from that LGTM!

**Comment by Miodec** [comment]
Just trying to figure out if its possible to add automatic hiding inside the ynab5 importer.

File: packages/loot-core/migrations/1685123304382_add_category_groups_hidden.sql [hunk 4]
```diff
@@ -0,0 +1,6 @@
+BEGIN TRANSACTION;
+
+
+ALTER TABLE category_groups ADD COLUMN hidden BOOLEAN NOT NULL DEFAULT 0;
```
[the remark this answers is not in the document]
Comment (Miodec) [reply packages/loot-core/migrations/1685123304382_add_category_groups_hidden.sql:4]: Done

**Review by j-f1 — APPROVED** [review]
Looks good to me, thank you!

**Review by MatissJanis — APPROVED** [review]
Amazing

**Comment by Kidglove57** [comment]
Loving the hidden categories feature. Working really well for me.

Thank you all for your hard work!

**Pull request merged**